File: src/main/java/xyz/chunkstories/api/item/inventory/InventoryTranslator.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.item.inventory;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import xyz.chunkstories.api.entity.Entity;
import xyz.chunkstories.api.entity.traits.Trait;
import xyz.chunkstories.api.entity.traits.serializable.TraitInventory;
import xyz.chunkstories.api.exceptions.world.WorldException;
import xyz.chunkstories.api.net.PacketReceptionContext;
import xyz.chunkstories.api.voxel.components.VoxelComponent;
import xyz.chunkstories.api.voxel.components.VoxelInventoryComponent;
import xyz.chunkstories.api.world.chunk.Chunk.ChunkCell;

import javax.annotation.Nullable;

/** Helper class to translate inventories accross hosts. The remote server has
 * no concept of an object reference from <b>your</b> heap, so you have to put a
 * little more work to describe what inventory you are talking about in terms
 * the both of you can understand.
 * 
 * <ul>
 * <li>For null inventories, you send a magic byte</li>
 *
 * <li>For Entity-relative inventories you send the entity UUID and the
 * component</li>
 *
 * <li>For Voxel-relative inventories, you send the xyz position and the
 * component id</li>
 *
 * <li>For other snowflake inventories made out of pixie dust, returns an
 * error</li>
 * </ul>
 */
public class InventoryTranslator {
	public static void writeInventoryHandle(DataOutputStream out, @Nullable Inventory inventory) throws IOException {
		/* if(inventory instanceof InventoryLocalCreativeMenu) out.writeByte(0x02);
		 * else */
		if (inventory == null || inventory.getHolder() == null)
			out.writeByte(0x00);
		else if (inventory instanceof TraitInventory) {
			TraitInventory entityInventory = (TraitInventory) inventory;

			out.writeByte(0x01);
			out.writeLong(((Entity) inventory.getHolder()).getUUID());
			out.writeShort(entityInventory.id());
		} else if (inventory.getHolder() instanceof VoxelInventoryComponent) {
			VoxelInventoryComponent component = (VoxelInventoryComponent) inventory.getHolder();

			out.writeByte(0x03);
			out.writeInt(component.getHolder().getX());
			out.writeInt(component.getHolder().getY());
			out.writeInt(component.getHolder().getZ());
			out.writeUTF(component.getName());
		} else
			out.writeByte(0x00);
		// throw new RuntimeException("Untranslatable and Unknown Inventory :
		// "+inventory+", can't describe it in outgoing packets");
	}

	@Nullable
	public static Inventory obtainInventoryHandle(DataInputStream in, PacketReceptionContext context) throws IOException {
		byte holderType = in.readByte();
		if (holderType == 0x01) {
			long uuid = in.readLong();
			short traitId = in.readShort();

			Entity entity = context.getWorld().getEntityByUUID(uuid);

			Trait trait = entity.traits.byId()[traitId];
			if (trait != null && trait instanceof TraitInventory) {
				return ((TraitInventory) trait);
			}
		} else if (holderType == 0x03) {
			int x = in.readInt();
			int y = in.readInt();
			int z = in.readInt();

			String traitName = in.readUTF();

			try {
				ChunkCell voxelContext = context.getWorld().peek(x, y, z);
				VoxelComponent com = voxelContext.components().getVoxelComponent(traitName);
				if (com != null && com instanceof VoxelInventoryComponent) {
					VoxelInventoryComponent comp = (VoxelInventoryComponent) com;
					return comp.getInventory();
				}
			} catch (WorldException e) {
				// TODO log as warning
				// Ignore and return null
			}
		}

		return null;
	}
}

File: src/main/java/xyz/chunkstories/api/item/inventory/InventoryTranslator.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.item.inventory

import xyz.chunkstories.api.entity.Entity
import xyz.chunkstories.api.entity.traits.serializable.TraitInventory
import xyz.chunkstories.api.exceptions.world.WorldException
import xyz.chunkstories.api.net.PacketReceptionContext
import xyz.chunkstories.api.voxel.components.VoxelComponent
import xyz.chunkstories.api.voxel.components.VoxelInventoryComponent
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/* Helper functions to translate inventories across hosts. The remote server has
 * no concept of an object reference from **your** heap, so you have to put a
 * little more work to describe what inventory you are talking about in terms
 * the both of you can understand.
 *
 *
 *  * For null inventories, you send a magic byte
 *
 *  * For Entity-relative inventories you send the entity UUID and the
 * component
 *
 *  * For Voxel-relative inventories, you send the xyz position and the
 * component id
 *
 *  * For other snowflake inventories made out of pixie dust, returns an
 * error
 *
 */

@Throws(IOException::class)
fun writeInventoryHandle(stream: DataOutputStream, inventory: Inventory?) {
    if (inventory == null) {
        stream.writeByte(0x00)
        return
    }

    when (val owner = inventory.owner) {
        null -> stream.writeByte(0x00)
        is Entity -> {
            //TraitInventory entityInventory = (TraitInventory) inventory;
            val trait = owner.traits.all().find { it is TraitInventory && it.inventory == inventory }
                    ?: throw Exception("This inventory is attached to an entity but that entity does not have a TraitInventory linking it back !")

            stream.writeByte(0x01)
            stream.writeLong((inventory.owner as Entity).UUID)
            stream.writeShort(trait.id)
        }
        is VoxelComponent -> {
            val component = inventory.owner as VoxelComponent?

            stream.writeByte(0x02)
            stream.writeInt(component!!.holder.x)
            stream.writeInt(component.holder.y)
            stream.writeInt(component.holder.z)
            stream.writeUTF(component.name)
        }
        else -> {
            stream.writeByte(0x00)
        }
    }
}

@Throws(IOException::class)
fun obtainInventoryByHandle(stream: DataInputStream, context: PacketReceptionContext): Inventory? {
    val holderType = stream.readByte()
    if (holderType.toInt() == 0x01) {
        val uuid = stream.readLong()
        val traitId = stream.readShort()
        val entity = context.world!!.getEntityByUUID(uuid)
        val trait = entity!!.traits.byId()[traitId.toInt()]

        if (trait is TraitInventory) {
            return trait.inventory
        }
    } else if (holderType.toInt() == 0x02) {
        val x = stream.readInt()
        val y = stream.readInt()
        val z = stream.readInt()

        val traitName = stream.readUTF()

        try {
            val voxelContext = context.world!!.tryPeek(x, y, z)
            val component = voxelContext.components.getVoxelComponent(traitName)
            if (component is VoxelInventoryComponent) {
                return component.inventory
            }
        } catch (e: WorldException) {
            // TODO log as warning
            // Ignore and return null
        }

    }

    return null
}

